### ISSUE TITLE
Meteor.loggingIn() waits for all subscriptions to finish before changing its state

### ISSUE THREAD

**Issue by ni-ko-o-kin** [meteor/meteor#10432]
Meteor: 1.8.2
OS: Ubuntu 19.04
Browser: Chrome 76, Firefox 71
reproduction repo: https://github.com/ni-ko-o-kin/meteor-loggingin-subscriptions

**The expected behavior:**
When a user logs in (with username: 'username' and password: 'password' in the reproduction-repo) Meteor.loggingIn() should return true and **after** it is finished the subscriptions should start to load.

In the reproduction-repo, after logging in, the page should first show 'logging in', then 'subs loading' and then the numbers of links.

**The actual behavior:**
After the user logs in, Meteor.loggingIn() return true **until** all subscriptions are loaded.

In the reproduction-repo, after logging in, the page shows 'logging in' until all subscriptions are loaded. 'subs loading' is never displayed.

**This only happens when the user logs in, but not when the user is already logged in and refreshes the page!**

You can login with `Meteor.loginWithPassword('username', 'password')`.

**Comment by ni-ko-o-kin** [comment]
Bug is still there with Meteor 1.8.2.

**Comment by filipenevola** [comment]
thanks @ni-ko-o-kin , I just labeled as `confirmed` so stale bot is going to ignore it.

**Comment by WilliamKelley** [comment]
I do not think this is a bug. This is canonical behavior that occurs because the publication depends on `this.userId` which is a reactive data source (setup internally with Tracker). 

When you use `Meteor.loginWith***` after the page loads, the client has already subscribed to the publication, which has been initialized for that connection and the dependency tracked. During a successful login, [`setUserId` is called](https://github.com/meteor/meteor/blob/devel/packages/accounts-base/accounts_server.js#L421). This can cause the publication to _immediately react (i.e. `Fibers.yield`)_, interrupting the login from returning, and start publishing those docs. Once the publication is done and returns, the server-side login finishes, returns to the client, and the client finishes, setting `loggingIn` to false. 

This is in contrast to when you reload the page; the `accounts-base` package runs [`_initLocalStorage`](https://github.com/meteor/meteor/blob/devel/packages/accounts-base/accounts_client.js#L535-L539) on load (before your component renders) and calls `loginWithToken`. In this case, the subscription _has not_ been initialized, and the dependency is _not_ tracked. Hence, there's no publication associated with the connection to interrupt the login process when the user id is set. Instead the login success returns to the client immediately. Only once this is done does your component get a chance to render and start the subscription.

The code was written in awareness of these facts:

- `accounts-client` noted that all subscriptions will be _rerun and ready_ by the time `loggingIn` will be set to `false`, [here](https://github.com/meteor/meteor/blob/devel/packages/accounts-base/accounts_client.js#L326-L329)
- `ddp-common` noted subscriptions will be _rerun_ in the definition of `setUserId` on the DDP MethodInvocation class, [here](https://github.com/meteor/meteor/blob/devel/packages/ddp-common/method_invocation.js#L47-L49)

---

As for a workaround, if you want to avoid this behavior for a given publication, you should condition your `subscribe` calls to only occur when the user is logged-in.

```js
let handle
if (Meteor.userId()) {
  handle = Meteor.subscribe('links.list');
} else {
  handle = undefined;
}
```
